### ISSUE TITLE
Fitting in with the Julia ecosystem

### ISSUE THREAD

**Issue by cscherrer** [JuliaStats/StatsBase.jl#116]
I'm working on a [BayesianLinearRegression.jl](https://github.com/cscherrer/BayesianLinearRegression.jl) module, and I'd like for it to fit in with the Julia ecosystem in a natural way. @johnmyleswhite has already been very helpful regarding general "idiomatic Julia" style, but it's still not clear to me how to best structure some things.

I guess I have two main questions:
* Is there a relatively simple module that show the "right way" to interface with formulas, the `StatisticalModel` abstraction, etc? The best option I've found so far for this is [GLM.jl](https://github.com/JuliaStats/GLM.jl), but if there's a simpler example, or even a guide for this, that would be very helful.
* Do you have any suggestions for implementing *Bayesian* models in particular? There's some intersection with the existing abstraction, but there are some values that are not represented here, such as model evidence.

Thanks!

**Comment by matthieugomez** [comment]
I'm wondering the same question (the first point). 

From what I understand,  the current advice is to define a new type of model and use it as the first argument to `fit`. 

But this leads to very different syntaxes across packages. Some packages like GLM use a type (not an instance of the type) as a first argument and add arguments to the `fit` function (like weight, tolerance, etc) while others (for instance NLreg) use an instance of the type as a first argument, and add arguments to the constructor function.  

In general, which option should be part of the constructor type and which should be part of the option for the fit method?  Is this question discussed somewhere?

This is the solution I like best:
- model-specific option should be specified as arguments to the constructor function, while options common across many models should be specified as arguments in the `fit` function ). This makes it natural for JuliaStats to define common abstract types for common options across different models (for instance e a dataframe, a formula, a weight vector, and a way to compute the standard errors).

For instance, fitting a fitting a linear model with `GLM` chould be written

```julia
fit(LinearModel(), y~x, df)
fit(LinearModel(maxIter=30), y~x, df)
```

Another issue is to define types that can contain symbols or vectors(if one wants to fit on matrix rather than a dataframe)

For instance, for the `offset` option in `glm`
```juli
fit(LinearModel(offset = z), y, X)
fit(LinearModel(offset = :z), y~x, df)
```

**Comment by nalimilan** [comment]
I don't really like the idea of separating arguments between `fit` and the constructor: what's the advantage? It sounds simpler to pass all arguments to `fit` methods, which will be specialized according to the model anyway. Else, people will always put the argument in the wrong place (for example, `maxIter`will likely be supported by many models, and yet it wouldn't be "standard" enough to be passed to `fit`).

**Comment by simonster** [comment]
I kind of see the argument for separating the arguments that control the model from the data. But using `LinearModel` as the constructor is a bit awkward because it has a type parameter that depends on the type of the data, which the user shouldn't have to specify when calling the constructor. Either the `LinearModel` passed to `fit` would be a `LinearModel{None}` (or something) and then `fit` would return a new `LinearModel{Float64}` (or whatever), or there would have to be two types: one that describes the model and fitting parameters, and one that represents the fit model. Overall I'm not totally sure the benefits of the constructor approach outweigh the awkwardness here.

In either case, I think `offset` should really be passed to `fit` and not the constructor, since it's really data and not part of the model description.

**Comment by nalimilan** [comment]
@dmbates What's your take on this? You wrote all of these packages, so you're the best person to say what are the pros and cons for each solution.

**Comment by matthieugomez** [comment]
Re-reading the code for [glm](https://github.com/JuliaStats/GLM.jl/blob/master/src/glmfit.jl), I actually don't understand the syntax of `fit` in GLM. 

There is a [`fit` method](https://github.com/JuliaStats/GLM.jl/blob/master/src/glmfit.jl#L200) that operates on a `Type{GeneralizedLinearModel}` (and is documented in the README). There is another [`fit` method](https://github.com/JuliaStats/GLM.jl/blob/master/src/glmfit.jl#L180) that operates on a `GeneralizedLinearModel` itself, and I'm not sure what this one does.

**Comment by nalimilan** [comment]
The second one just looks like an alternative interface, which can be used to re-fit a model with different parameters or data, but without recomputing the whole design matrix? But since it appears to modify the object in place, it would better be called `fit!`.

**Comment by simonster** [comment]
I think I wrote that second `fit` method. It refits an existing model object using a new response vector. I think I wrote it before StatsBase included `fit!`, but now it should be renamed (also documented).

**Comment by bicycle1885** [comment]
This may be off the topic, but let me know the convention of the "design matrix" in Julia.
The document states that a model can be [`fit`ted](http://statsbasejl.readthedocs.org/en/latest/statmodels.html#fit) using a design matrix, but the layout of observations is unclear. Should observations be along with columns or rows?
Since Julia has column-major matrices, I think columns should be observations in terms of performance.
[MNIST.jl](https://github.com/johnmyleswhite/MNIST.jl) adopts this layout.

I want to know your opinions and want to see written document on the choice for later reference.

**Comment by johnmyleswhite** [comment]
I would like us to standardize on using column-major layout everywhere, but that would require some changes to GLM, IIRC. Sadly, we lack the kind of BDFL that we need to make progress on these problems.

**Comment by andreasnoack** [comment]
@bicycle1885 I think the preferences here depend on background and how you typically fit your models. @lindahua and @johnmyleswhite have a preference for treating columns as observations, so in packages where their are the main contributors, you will probably see observations by column.

@dmbates, who wrote GLM.jl, has followed, what I consider, the statistics tradition and assumes that rows are observations. All design matrices I've seen in school followed that tradition and I'm pretty sure it is the more efficient choice for regression problems. As you mention, it is probably not the case in other types of models.

So the choice of observations as rows vs. columns is more dependent on who is most actively developing a specific package than a general consensus. It would be great if we could find some kind of consensus here, but so far we haven't.

**Comment by johnmyleswhite** [comment]
I have some hope we can make a little progress if transpose becomes a lazy operation.

@andreasnoack: I agree that the presentation in statistics is always to use rows as observations, but I'm not aware of any performance gains from using this encoding. Do you have examples?

**Comment by nalimilan** [comment]
Conversely, what's the performance gain in putting observations in columns?

**Comment by andreasnoack** [comment]
It's mainly based on the observation that for least squares you QR the design matrix if the observations are by row and you LQ it if observations are by column. Im pretty sure forming and applying reflectors is faster for QR than LQ when the storage is column major.

OpenBLAS:
```jl
julia> A = randn(10000,100);

julia> At = A';

julia> @time LAPACK.geqrf!(copy(A));
  0.029051 seconds (13 allocations: 7.655 MB, 22.11% gc time)

julia> @time LAPACK.gelqf!(copy(At));
  0.274455 seconds (14 allocations: 7.655 MB)
```

MKL:
```jl
julia> At = A';

julia> @time LAPACK.geqrf!(copy(A));
  0.028761 seconds (13 allocations: 7.715 MB)

julia> @time LAPACK.gelqf!(copy(At));
  0.122216 seconds (14 allocations: 7.655 MB)
```

**Comment by dmbates** [comment]
The big difference between observations as rows and observations as columns occurs in the data frame representation where the columns can be heterogeneous.  A categorical covariate has a different representation than a numerical covariate which has a different representation from a date/time, ...

The most natural way of creating a model matrix is by operating on columns of a data frame and generating the model matrix column by column.

As @andreasnoack points out, statisticians tend to think of a model matrix with rows as observations and to use numerical methods like the QR decomposition in least squares calculations.  In situations where you explicitly form the product X'X for the observations as rows formulation or XX' for the columns as observations, then use a Cholesky factorization it doesn't really matter how you store the model matrix.  I end up using both under different circumstances.

**Comment by johnmyleswhite** [comment]
@nalimilan: The performance gain for many algorithms when observations are columns comes from fewer cache misses since the columns are contiguous in memory, but the rows are not. You can see this even in something as simple as comparing `mean(A, 1)` with `mean(A, 2)` when `A = randn(25_000, 25_000)`. On my machine, the row means are roughly 50% slower to compute.

@dmbates: What I don't think I understand is why we can't take a DataFrame where rows are observations and generate a model matrix where columns are observations.That approach seems to me like the path of least resistance, especially in light of your suggestion that most GLM fitting depends upon the scatter matrix, which is symmetric.

I would love to see some progress on this issue. Maybe we can get @malmaud to unify the tribes?

**Comment by dmbates** [comment]
@johnmyleswhite In column-major ordering it is easier to append columns than to append rows.  If you build up the model matrix a term at a time you would be doing more expensive operations if columns are observations.  Of course, profiling this would probably show very little difference between on orientation and the other.

**Comment by nalimilan** [comment]
@johnmyleswhite I know this, but what I don't understand is why you would like to work on a row-by-row basis. For example, taking the mean is usually done per variable, not for different variables of one observation, in which case having variables in columns is more efficient.

**Comment by johnmyleswhite** [comment]
Because you usually need to do most operations in terms of observations, not features? Consider something like finding a nearest neighbor. You're going to iterate over observations, comparing each with a chosen exemplar.

**Comment by simonster** [comment]
But for something like coordinate descent you really do want to do operations in terms of features. Picking the convention based on performance seems a bit strange when we know there is no single answer that applies for all algorithms.

**Comment by nalimilan** [comment]
That's what I was going to say. I'd rather follow the convention used everywhere in stats, since performance seems to depend on the use case.

With lazy transpose support, wouldn't this issue disappear anyway?

**Comment by johnmyleswhite** [comment]
@simonster: I agree that the performance gap is probably small enough that we'd be better off just resolving the issue once and for all rather than worrying about <2x perf wins.

@nalimilan: "Used everywhere in stats" is effectively a restatement of the external-to-Julia cultural disagreements that have created the problems we already have. Lots of people come to Julia from both machine learning and Matlab, where the column=observation tradition is strong.

Lazy transpose support helps a bit, but doesn't fully solve the issue since you still need to make assumptions about canonical orientations before applying any lazy transposes. It solves some of the perf issues, but it doesn't solve the problem that you need to be able to safely assume in all code that people are providing functions with arguments that satisfy certain assumptions.

**Comment by simonster** [comment]
I would prefer row=observation as well, but this is one of those cases where any convention is better than no convention. In MATLAB I don't think there really is a convention. Certainly `linreg` and `glmfit` follow row=observation.

**Comment by johnmyleswhite** [comment]
Can we get @lindahua and @timholy to agree to the row=observation convention?

**Comment by lindahua** [comment]
I mainly come from a machine learning background and currently working on research involving huge data sets. 

In machine learning, we generally deal with *features* (e.g. a 1000-dimensional numeric vector obtained from CNN to encode an image), instead of things like a data frame. So generally, we scan the data set observation by observation. For such cases, using columns as observations do have performance benefits. 

A more important argument in favor of observations-as-columns style is that we are facing even larger datasets, where samples come in as a stream. At each iteration, we take a subset of observations, called a *mini-batch*, and do some processing. 

I certainly understand why observations-as-rows are preferable in the statistics community -- however, there are other *very important* contexts, it is very desirable to have the data organized in such a way that elements of an observation are placed together.

**Comment by lindahua** [comment]
I think we have to come up something to cater to both scenarios (data-frame-like data in statistics and observation-set in machine learning).

**Comment by johnmyleswhite** [comment]
Why can't a minibatch represent the data in the form of rows-as-observations? It's not clear to me how streaming data support has any intrinsic relationship to columns-as-observations.

**Comment by lindahua** [comment]
Conceptually, it works either way. Performance-wise, they can be very different. 

In my mind, this is not actually not a column/row problem. To have better performance, we want to store things by observations (both in memory and in the disk), first sample, followed by the second sample, followed by the third sample, and so on ...

In this way, when we want to get one sample, we don't have to jump to 1000 places that are far away from each other.

**Comment by lindahua** [comment]
When designing a data structure, I always have such a scenario in my mind:

I have a data set comprised of 1000000 images, each encoded into a 1000-dimensional vector, and I want to do the following:

```julia
for sample in dataset
    # ... do something (e.g. compute the gradient and update the param)
end    

```

**Comment by dmbates** [comment]
Another argument for columns-as-observations is that Distributed Arrays or Shared Arrays split more naturally in that orientation.  Generally you want contiguous storage in the chunks and you also want to work by observation.

It seems that it would be worthwhile allowing for either orientation in the basic algorithms.

**Comment by johnmyleswhite** [comment]
I'm ok with that, but then we should find a way to standardize on an input argument that all functions use to indicate the orientation.

**Comment by Evizero** [comment]
Even though I am also used to the notation of having rows represent observations, I am currently following the convention of implementing everything in terms of columns denoting observations. I have done some very informal benchmarks for myself in the beginning (based on the suggestion in the performance tips section of the julia docs) and they clearly favoured this approach. The ML algorithms that I currently work with (e.g. for SVMs) seem to profit from that choice because I usually end up lopping over the observations using ArrayViews into the individual columns of the design matrix (i.e. I also loop over the samples in the dataset - even for dual coordinate descent btw). 

I too actually prefer the row-based notation conceptually, but unless a consensus for a standard is reached by the community I am unlikely to change to it. For now I think I'll stick to columns as observations simply because it currently yields better performance for my problems

**Comment by cstjean** [comment]
> I don't really like the idea of separating arguments between fit and the constructor: what's the advantage?

When the model is specified in the constructor, it's easy to fit the same model on multiple datasets, in particular for cross-validation. Scikit-learn does it this way, and I find it generally more pleasing than passing a closure, as is done with MLBase.

Re. conventions: I think there's a great opportunity here to outdo Scikit-learn with a _decentralized_ set of models, implemented across multiple packages. But someone needs to make clear decisions about what the [API](http://statsbasejl.readthedocs.org/en/latest/statmodels.html) should be and publicize it for others to implement. IMO it's more important to get the ball rolling in a consistent direction than to get the details perfectly right.

**Comment by nalimilan** [comment]
@cstjean But if some arguments are not part of the constructor, then you'll have to pass them again when refitting the model on a different dataset. You won't be able to run it again with exactly the same settings only from a saved model object. That's why I'd prefer all model parameters to be specified in the same place, and be saved in the model; only the data would be passed separately to `fit`.

**Comment by cstjean** [comment]
> But if some arguments are not part of the constructor, then you'll have to pass them again when refitting the model on a different dataset. 

In scikit-learn, all arguments are part of the constructor except the data. 

> That's why I'd prefer all model parameters to be specified in the same place, and be saved in the model; only the data would be passed separately to fit.

I don't follow, could you give an example? It sounds to me like we're suggesting the same thing :)

In sklearn, you can have something like:

```julia
model = Pipeline(StandardScaler(), PCA(n_components=100), RidgeRegression(regularization=0.3)
fit!(model, X, y)
```

I'm genuinely curious how other frameworks solve this data pipeline issue. I don't see a clean way to do it with `fit!(::ModelType, X, y, modelparameters...)`, but perhaps I just lack imagination!

**Comment by StefanKarpinski** [comment]
> In scikit-learn, all arguments are part of the constructor except the data.

@amueller: do you find this design choice to have been a good one or does it have awkward cases?

**Comment by amueller** [comment]
[I haven't read this threat yet]
The API works well in most cases.
One awkward case is if the parameters are actually more parameters of the prediction than the fitting part. For example, say you use feature selection using some threshold. Then, it is silly to refit the whole model, just to set the threshold.

What you can do at the moment in scikit-learn is that for the threshold thing, you could do (this is not actually an existing class):
```python
feature_selection = FeatureSelection(threshold=0.2)
feature_selection.fit(X, y)
transformation_02 = feature_selection.transform(X)
feature_selection.threshold = 0.1
transformation_01 = feature_selection.transform(X)
```
The problem is that changing a parameter after fitting will not have an effect for most parameters.
From the API (and unfortunately also the docs at the moment) it is unclear for which parameters a change after ``fit`` has an effect.

**Comment by amueller** [comment]
Another somewhat awkward case is that it is sometimes not entirely clear what is data and what is parameters. For example, we have ``class_weights`` that say how important each class is, and we have ``sample_weights`` that say how important each sample is. ``class_weights`` is a parameter of the model, specified in the constructor. ``sample_weights`` is part of the data from an API standpoint (and gave us quite some headaches).
Basically, everything that is indexed by samples is part of the data. The reason is the way we do cross-validation. The model doesn't know which chunk of the data it got passed. So if we gave the ``sample_weights`` at construction time, there would be no way to match it up with the samples it gets during training.

There could be other things that are per-sample, like group ids for cross-validation.

**Comment by amueller** [comment]
> I think there's a great opportunity here to outdo Scikit-learn with a decentralized library of models, implemented across multiple packages.

This has pros and cons. We certainly struggle with our role sometimes. It is really hard to decide what is within scope. And many specialized algorithms are not included. On the other hand, people could totally build an eco-system **around** scikit-learn. That has not really happened so far, in particular because there is no good platform for this (pypi is too large to discover anything maybe?).
People have been proposing scikit-learn incubators, or a contrib repo, but nothing has really manifested itself.

R has the distributed model, but it basically has no shared infrastructure or interface. Carret needs to wrap everything, which is clearly not what you want. On the other hand, CRAN takes away the need for one centralized place. However, that creates often multiple solutions to the same problem, but with slightly different interfaces (multiple ways to compute a ROC curve for example).

Some people also told me that they like the way that models inside scikit-learn are "vetted" by the core-developers. They trust code in scikit-learn more than they would trust a random package. [Whether that is reasonable or not is another question.]

**Comment by cstjean** [comment]
@amueller Thank you for your comments.

> Some people also told me that they like the way that models inside scikit-learn are "vetted" by the core-developers. They trust code in scikit-learn more than they would trust a random package. [Whether that is reasonable or not is another question.]

That's a very good point about the value of curation.

I have a ton of free time for the next few months, and I've been thinking about implementing the scikit-learn interface for Julia libraries, as well as wrappers around the Python scikit-learn models (similar to Orchestra.jl). In the short term, it would help us avoid arguments about which conventions are best, and it doesn't preclude defining another API once we have a clearer perspective on the trade-offs involved.

Julia "bootstrapped" off of Matlab's syntax, we can apply the same strategy for ML.

**Comment by StefanKarpinski** [comment]
> Some people also told me that they like the way that models inside scikit-learn are "vetted" by the core-developers. They trust code in scikit-learn more than they would trust a random package. [Whether that is reasonable or not is another question.]

To me this seems like one of scikit-learn's great strengths: coherent interfaces and consistent quality.

**Comment by StefanKarpinski** [comment]
Thanks for the feedback, @amueller!

**Comment by nalimilan** [comment]
@cstjean Yes, we're probably proposing the same thing. Even though @amueller showed the distinction between model specification and data isn't always easy to do, trying to stick to this rule is already a good start. This distinction goes against what @matthieugomez proposed above (https://github.com/JuliaStats/StatsBase.jl/issues/116#issuecomment-114226641).

**Comment by datnamer** [comment]
@cstjean, @tbreloff and @Evizero  have already started planning and executing some ML infrastructure. I would coordinate with them so we avoid duplicating efforts.

Also did everyone see @simonbyrne 's Julia Stats roadmap here? : https://groups.google.com/forum/#!topic/julia-stats/29l5yA87Qss 

Perhaps we can also coordinate with these efforts. Do we want to combine stats and ML or have them separate areas? 

I think there should be as much reuse of algorithms, infrastructure and interface as possible. Especially since Julia's awesome autodiff and thing like [https://lendle.github.io/TargetedLearning.jl/](Targeted Learning) allow for the derivation of statistical parameters from stochastic machine learning, IIRC. 

Regarding degree of centralization, no reason we can't have the best of both worlds: A core of vetted and widely useful models with interface to plug in third party packages and CRAN like task views to incentivize  ecosystem growth. 

As far as API and first steps, I think Julia ML/Stats should pull a JuliaOpt and focus on a general model based machine learning framework using streaming and distributed variational inference. This is a niche Julia lends itself to be best in breed and is widely applicable to different problems. I think this would pull more users than trying to improvie recreate sklearn (which should be a goal as well).

**Comment by Evizero** [comment]
Since I was pinged: I am still very interested in this. That being said, I will transition jobs soon. This means I am in the middle of tidying up the projects I am currently in charge of. If everything goes smoothly I'll be able to devote time to JuliaML again in a few weeks. Until then I am afraid I am of little help. I do try to keep up with reading the comments though.

**Comment by ValdarT** [comment]
Is it perhaps time to revisit this? 
Repeating my proposal from [Discourse](https://discourse.julialang.org/t/machine-learning-toolset-improvement/4835/8), perhaps it would be good to move the modelling abstractions from `StatsBase.jl` to `StatsModels.jl` and redefine/extend them so that they would meet the needs of both JuliaStats org as well as `ScikitLearn.jl`. This way there would be one obvious abstraction to extend and Julia users could use the pure-Julia algorithms maintained by JuliaStats org with the familiar and "battle-tested" ScikitLearn interface. Or perhaps there's a better idea.
In any case, seems that the ecosystem could be considerably improved with not too much effort.

**Comment by datnamer** [comment]
CC @oxinabox and @cstjean  who might be interested.

**Comment by cstjean** [comment]
I'm all for simplifying the ecosystem, but one would have to build a concrete proposal that deals with all the differences between the interfaces. For instance, hyperparameters are passed to the model constructor in ScikitLearn.jl, but IIRC in StatsBase, they are passed to `fit`.

**Comment by ararslan** [comment]
Moving the model abstractions to StatsModels has been on the agenda for a very long time, someone just needs to do it. Once the Statistics stdlib package can be developed separately from Julia itself I think it will be a good time to do a mass migration of functionality to different packages, including Statistics (for e.g. weights) and StatsModels (modeling abstractions).
